cli/validation/validate.go

```diff
@@ -36,26 +43,53 @@ func ValidateDeployment(name string, policyPath string, project string) (validat
 			return false, err
 		}
 
-		if mapper, ok := mappers[res.Type]; ok {
-			asset, err := mapper(res.Type, res)
-			if err != nil {
-				log.Printf("error getting resource: %s, deployment: %s, error: %v", res.Name, name, err)
-				return false, err
-			}
-			asset.Ancestry, err = getAncestry(project)
-			if err != nil {
-				log.Printf("error getting resource ancestry: %s, deployment: %s, error: %v", res.Name, name, err)
-				return false, err
-			}
+		if mapperList, ok := mappers[res.Type]; ok {
+			for _, mapper := range mapperList {
+				asset, err := mapper.convert(res.Type, res)
+				if err != nil {
+					log.Printf("error getting resource: %s, deployment: %s, error: %v", res.Name, name, err)
+					return false, err
+				}
+				asset.Ancestry, err = getAncestry(project)
+				if err != nil {
+					log.Printf("error getting resource ancestry: %s, deployment: %s, error: %v", res.Name, name, err)
+					return false, err
+				}
+
+				key := asset.Type + "__" + asset.Name
+
+				if existing, exists := assets[key]; exists {
+					// The existance of a merge function signals that this tf resource maps to a
+					// patching operation on an API resource.
+					if mapper.merge != nil {
+						asset, err = mapper.merge(existing, asset)
+						if err != nil {
+							return false, err
+						}
+					} else {
+						log.Printf("duplicate asset: %s, deployment: %s, error: %v", key, name, err)
+						return false, errors.New(fmt.Sprintf("duplicate asset: %s", key))
+					}
+				}
 
-			assets = append(assets, asset)
+				assets[key] = asset
+			}
 		} else {
 			log.Printf("resource type not supported: %s, name %s, deployment: %s, project %s\n", res.Type, res.Name, name, project)
 		}
 	}
 
+	getAssets := func(assets map[string]cai.Asset) byName {
```

Comment:
It's pretty inefficient to rebuild this entire list at the end. Instead, I think we should be building the master slice from the beginning. We can, in parallel, also maintain a map() of names to asset indexes in the master list.